Organize and Update "Local Functions vs Lambda Expressions" section in Local Functions page

jmarkman opened the pull request:
## Summary

The original issue author found that there wasn't enough separation in the topics covered in the contrast section for local functions and lambda expressions. There were enough H2 (`##`) for readers to jump around the Local Functions page, but I added several H3s (`###`) for each facet being contrasted.  As it exists now, the contrast section is one large chunk of information. My H3 changes broke that large chunk apart so it didn't feel like a never-ending information dump and was easier to navigate. 

I also adjusted some of the wording to reflect the new division of content. As the original author of the issue stated, there's a rogue sentence something along the lines of "Third, the compiler can do [blah]" when there was no previous counting in the section. No reason to go back and add counting now that the facets are outlined.

I also added a very barebones example for `yield return` for local functions. While there's a `yield return` example in the "Local Functions and Exceptions" section, it was purposefully made to demonstrate using exceptions, so I felt that a separate snippet would be clearer in purpose. I also added a link to the specific compiler error that would appear if a developer tried to use `yield return` with a lambda expression. I don't know the stance on media in the docs, so I was averse to adding an image depicting what a developer would see in Visual Studio when attempting such a feat.

I do have a question, however. I wanted to add a link to the definite assignment section in the [language spec page for variables](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/variables#definite-assignment) to the subsection where definite assignment in regards to local functions/lambda expressions is brought up, but I haven't been able to figure out the proper "syntax" for the markdown URL. Would I just copy the href for the Variables page from `toc.yml` and append the `#definite-assignment` section link to it and use *that* as the URL?

As always, please let me know if I've interpreted something incorrectly, something doesn't sound right, something could be better, etc., and I'll do my best to fix it.

Fixes #16426

jmarkman committed:
improvements to contrast section
- use h3 headings to highlight differences
- add spacing between thoughts so semi-related expressions
don't read as if they were run-on sentences
- add example for yield return for local function
- add link to compiler error in yield return section

jmarkman committed:
Merge branch 'master' of https://github.com/dotnet/docs into jmarkman-organize-local-funcs-page

jmarkman committed:
refactor to use new code snippet insertion pattern
- remove YieldLocalFunc file
- add method to snippets\local-functions\Program.cs
- update syntax in local-functions.md

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+Local functions are explicitly named like methods. Lambda expressions are anonymous methods and need to be assigned to variables that are either `Func` or `Action` types. When you declare a local function, the process is like writing a normal method; you declare a return type and a function signature.
 
-Definite assignment rules also affect any variables that are captured by the local function or lambda expression. Both local functions and lambda expression rules demand that any captured variables are definitely assigned at the point when the local function or lambda expression is converted to a delegate. The difference is that lambda expressions are converted to delegates when they are declared ...
+### Typing
```
Guidelines recommend to not use "ing"-words.

BillWagner replied:
I'll suggest the following change:

```suggestion
### Function signatures and lambda expression types
```

jmarkman replied:
👍

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
-Third, the compiler can perform static analysis that enables local functions to definitely assign captured variables in the enclosing scope. Consider this example:
+Lambda expressions rely on the typing of the `Func`/`Action` variable that they're assigned to determine the argument and return types. In local functions, since the syntax is much like writing a normal method, argument types and return type are already part of the function declaration.
+
+### Definite Assignment
```
Titles should be sentence-cased:
```suggestion
### Definite assignment
```

jmarkman replied:
Will sentence-case the titles that you pointed out

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+
+These differences mean that recursive algorithms are easier to create using local functions. You can declare and define a local function that calls itself. Lambda expressions must be declared, and assigned a default value before they can be re-assigned to a body that references the same lambda expression.
+
+### Conversion to a Delegate
```
```suggestion
### Conversion to a delegate
```

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+
+If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate. That rule enables you to declare a local function at any convenient location in its enclosing scope. It's common to declare local functions at the end of the parent method, after any return statements.
+
+### Variable Capture
```
```suggestion
### Variable capture
```

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -110,7 +127,9 @@ int M()
 The compiler can determine that `LocalFunction` definitely assigns `y` when called. Because `LocalFunction` is called before the `return` statement, `y` is definitely assigned at the `return` statement.
 
-The analysis that enables the example analysis enables the fourth difference. Depending on their use, local functions can avoid heap allocations that are always necessary for lambda expressions. If a local function is never converted to a delegate, and none of the variables captured by the local function is captured by other lambdas or local functions that are converted to delegates, the compiler ...
+### Heap Allocations
```
```suggestion
### Heap allocations
```

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/snippets/local-functions/Program.cs:
```diff
@@ -121,5 +121,27 @@ private async Task<int> SecondStep(int index, string name)
         }
+
+        //<YieldReturn>
+        public IEnumerable<string> SequenceToLowercase(IEnumerable<string> input)
```
The project doesn't build now because of the missing `using` directive. Please add `using System.Collections.Generic;` to the top of the file.

jmarkman replied:
Ah! I'm embarrassed 😞 - I'll fix this first

I'm using Visual Studio Code to make my changes; is there a way to run a local build to ensure I'm not forgetting anything that would break the actual build in the future?

BillWagner replied:
Since this code is inline, there isn't. 😢  

We've been working toward using all code includes, but we haven't completed that transition yet. Where those are in place, all the displayed code runs through our CI system.

pkulikov replied:
@jmarkman just build the project in the corresponding folder: `docs/csharp/programming-guide/classes-and-structs/snippets/local-functions`

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/snippets/local-functions/Program.cs:
```diff
@@ -121,5 +121,27 @@ private async Task<int> SecondStep(int index, string name)
+        //<YieldReturn>
+        public IEnumerable<string> SequenceToLowercase(IEnumerable<string> input)
+        {
+            if (input.Count() == 0)
```
```suggestion
            if (!input.Any())
```

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/snippets/local-functions/Program.cs:
```diff
@@ -121,5 +121,27 @@ private async Task<int> SecondStep(int index, string name)
+        {
+            if (input.Count() == 0)
+            {
+                throw new Exception("There are no items to convert to lowercase!");
```
Throwing bare exceptions is not recommended. In this context it looks to be `ArgumentException`
```suggestion
                throw new ArgumentException("There are no items to convert to lowercase.");
```

jmarkman replied:
Normally I don't throw base exceptions, but my assumption was that the `yield return` was the focus of the demonstration. I'll replace it with `ArgumentException`

pkulikov commented on docs/csharp/programming-guide/classes-and-structs/snippets/local-functions/Program.cs:
```diff
@@ -121,5 +121,27 @@ private async Task<int> SecondStep(int index, string name)
+            {
+                foreach (var item in input)
+                {
+                    var output = item.ToLower();
+                    yield return output;
```
```suggestion
                    yield return item.ToLower();
```

pkulikov reviewed (commented):
Nit-picks on text and code style

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+
+Local functions have different rules for definite assignment than lambda expressions. A local function declaration can be referenced from any code location where it is in scope; in our first example `LocalFunctionFactorial`, we could declare our local function either above or below the `return` statement and not trigger any compiler errors.
+
+Lambda expressions follow the rules for definite assignment more closely. In order for a lambda expression to be used, the `Func`/`Action` variable that it will be assigned to must be declared and the lambda expression assigned to it. Notice that `LambdaFactorial` must declare and initialize the lambda expression `nthFactorial` before defining it. Not doing so results in a compile time error for  ...
```
Related to previous suggestion:

```suggestion
Lambda expressions are objects that are assigned at runtime. Lambda expressions follow the rules for definite assignment. In order for a lambda expression to be used, the `Func`/`Action` variable that it will be assigned to must be declared and the lambda expression assigned to it. Notice that `LambdaFactorial` must declare and initialize the lambda expression `nthFactorial` before defining it. Not doing so results in a compile time error for referencing `nthFactorial` before assigning it.
```

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+
+### Conversion to a Delegate
+
+Both local functions and lambda expressions require that any captured variables are definitely assigned when either is converted to a delegate, but *when* they're converted differs. Lambda expressions are converted to delegates when they're declared, while local functions are only converted to delegates when ***used*** as a delegate.
+
+If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate. That rule enables you to declare a local function at any convenient location in its enclosing scope. It's common to declare local functions at the end of the parent method, after any return statements.
```
These two paragraphs have a number of subtle and misleading statements.  As written, I don't have a great suggestion, but here are the issues with it. 

- Its mixing a discussion of the definite assignment rules for captured variables and conversion to a delegate. 
- It's an implementation detail whether or not a local function is implemented as a delegate. (When it happens, the delegate definition is a compiler generated name with private scope, and can't be referenced as such from source code). 
- The sentence on declaration location is covered better above where you make the distinction between a function definition and a lambda declaration and assignment.

I think this H3 can be removed. If you chose that route, I'd add a sentence in the next section on variable capture. Namely, that local functions are implemented as delegate types when they capture variables in the enclosing scope.

jmarkman replied:
I've been sitting on this part for a while. This was the hardest part of the section to parse when I was reading the current live docs since I believe there were multiple topics being discussed in the same paragraph. I *think* having the H3 is beneficial so long as I'm accurately covering what you've specified, so I'm toying with the following:

---

### Definite assignment

Lambda expressions are objects that are declared and assigned at runtime. In order for a lambda expression to be used, it needs to be [definitely assigned](../../../../_csharplang/spec/variables.md#definite-assignment): the `Func`/`Action` variable that it will be assigned to must be declared and the lambda expression assigned to it. Notice that `LambdaFactorial` must declare and initialize the lambda expression `nthFactorial` before defining it. Not doing so results in a compile time error for referencing `nthFactorial` before assigning it.

Local functions are defined at compile time. As they're not assigned to variables, they can be referenced from any code location **where it is in scope**; in our first example `LocalFunctionFactorial`, we could declare our local function either above or below the `return` statement and not trigger any compiler errors.

These differences mean that recursive algorithms are easier to create using local functions. You can declare and define a local function that calls itself. Lambda expressions must be declared, and assigned a default value before they can be re-assigned to a body that references the same lambda expression.

### Variable capture

The rules of [definite assignment](../../../../_csharplang/spec/variables.md#definite-assignment) also affect any variables that are captured by the local function or lambda expression. The compiler can perform static analysis that enables local functions to definitely assign captured variables in the enclosing scope. Consider this example:

```csharp
// code
```

The compiler can determine that `LocalFunction` definitely assigns `y` when called. Because `LocalFunction` is called before the `return` statement, `y` is definitely assigned at the `return` statement.

Note that when a local function captures variables in the enclosing scope, the local function is implemented as a delegate type in the eyes of the compiler.

---

The trailing sentence on declaration location is removed and there's no mixing of definite assignment and delegate conversion. However, I'm not entirely sure how to handle the delegate conversion part since it's an implementation detail. This comment is already getting pretty long but I'm not sure if the following should stay as its own section or if it should be moved into the `Variable capture` section:

---

### Implementation as a delegate

Lambda expressions are converted to delegates when they're declared. Local functions are more flexible in that they have the *option* to be implemented as a delegate. Local functions are only converted to delegates when they're used as delegates. If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate.

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -110,7 +127,9 @@ int M()
-The analysis that enables the example analysis enables the fourth difference. Depending on their use, local functions can avoid heap allocations that are always necessary for lambda expressions. If a local function is never converted to a delegate, and none of the variables captured by the local function is captured by other lambdas or local functions that are converted to delegates, the compiler ...
+### Heap Allocations
+
+Depending on their use, local functions can avoid heap allocations that are always necessary for lambda expressions. If a local function is never converted to a delegate, and none of the variables captured by the local function is captured by other lambdas or local functions that are converted to delegates, the compiler can avoid heap allocations.
```
I really like this paragraph. I'd add that if you want to guarantee that behavior, you can declare a `static` local function.

BillWagner reviewed (changes requested):
Thanks for making these changes @jmarkman There's a lot of great work here. I also really appreciate the detailed description on the PR. That's super helpful.

I had some detailed comments in a couple areas. These areas are confusing to readers, and I think the extra clarifications will help.

Once you make these updates, I'll give it another look. If you have any questions, just @ mention me.

jmarkman commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+### Typing
 
-Third, the compiler can perform static analysis that enables local functions to definitely assign captured variables in the enclosing scope. Consider this example:
+Lambda expressions rely on the typing of the `Func`/`Action` variable that they're assigned to determine the argument and return types. In local functions, since the syntax is much like writing a normal method, argument types and return type are already part of the function declaration.
```
@BillWagner: after changing typing -> type, should I change the line to use "`delegate` variable" instead of "`Func`/`Action` variable"? I want to keep it consistent with the change you suggested for the "Naming" section, but I'm not sure if having it read 
> "Lambda expressions rely on the type of the `delegate` variable that they're assigned to..." 

alone is fine, or have the sentence look like

> "Lambda expressions rely on the type of the `delegate` variable (again, typically `Func` or `Action` types) that they're assigned to..."

to repeat that these are usually `Func`/`Action` types, *or* just leave it as it is with the typing -> type change.

BillWagner replied:
Great question. When I read it the first time, I think it's fine to use `Func` / `Action` everywhere after the first use. In practice, those types are preferred over declaring a delegate type. The change from "typing" to "type" was all that I thought was necessary.

IEvangelist commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+### Naming
 
-Local functions have different rules for definite assignment than lambda expressions. A local function declaration can be referenced from any code location where it is in scope. A lambda expression must be assigned to a delegate variable before it can be accessed (or called through the delegate referencing the lambda expression). Notice that the version using the lambda expression must declare an ...
+Local functions are explicitly named like methods. Lambda expressions are anonymous methods and need to be assigned to variables that are either `Func` or `Action` types. When you declare a local function, the process is like writing a normal method; you declare a return type and a function signature.
```
```suggestion
Local functions are explicitly named like methods. Lambda expressions are anonymous methods and need to be assigned to variables that are either `Action` or `Func` types. When you declare a local function, the process is like writing a normal method; you declare a return type and a function signature.
```

Order these by type alphabetically, and most appropriate in terms of complexity - since `Func` has return types other than `void`.

IEvangelist commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+### Typing
 
-Third, the compiler can perform static analysis that enables local functions to definitely assign captured variables in the enclosing scope. Consider this example:
+Lambda expressions rely on the typing of the `Func`/`Action` variable that they're assigned to determine the argument and return types. In local functions, since the syntax is much like writing a normal method, argument types and return type are already part of the function declaration.
```
```suggestion
Lambda expressions rely on the typing of the `Action` or `Func` variable that they're assigned to determine the argument and return types. In local functions, since the syntax is much like writing a normal method, argument types and return type are already part of the function declaration.
```

IEvangelist commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -80,22 +80,39 @@ Like with the method iterator, you can refactor the preceding example and put th
+
+Both local functions and lambda expressions require that any captured variables are definitely assigned when either is converted to a delegate, but *when* they're converted differs. Lambda expressions are converted to delegates when they're declared, while local functions are only converted to delegates when ***used*** as a delegate.
+
+If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate. That rule enables you to declare a local function at any convenient location in its enclosing scope. It's common to declare local functions at the end of the parent method, after any return statements.
```
```suggestion
If you declare a local function and only reference it by calling it like a method, it will not be converted to a delegate. That rule enables you to declare a local function at any convenient location in its enclosing scope. It's common to declare local functions at the end of the parent method, even after return statements.
```

I prefer phrasing it this way as it is often alarming to people that these can be declared after `return` statements.

IEvangelist commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -125,7 +144,13 @@ The instantiation necessary for lambda expressions means extra memory allocation
 :::code language="csharp" source="snippets/local-functions/Program.cs" id="AsyncWithLocal" :::
 
-One final advantage not demonstrated in this sample is that local functions can be implemented as iterators, using the `yield return` syntax to produce a sequence of values. The `yield return` statement is not allowed in lambda expressions.
+### Usage of the `yield` Keyword
```
```suggestion
### Usage of the `yield` keyword
```

IEvangelist commented on docs/csharp/programming-guide/classes-and-structs/snippets/local-functions/Program.cs:
```diff
@@ -121,5 +121,27 @@ private async Task<int> SecondStep(int index, string name)
+            
+            IEnumerable<string> LowercaseIterator()
+            {
+                foreach (var item in input)
+                {
+                    var output = item.ToLower();
+                    yield return output;
```
```suggestion
                foreach (var output in input.Select(item => item.ToLower()))
                {
                    yield return output;
```

Would require `using System.Linq;` though, so go with @pkulikov [suggestion instead](https://github.com/dotnet/docs/pull/21117/files#r505227879).

jmarkman replied:
I think I could do this since @pkulikov suggested using `if (input.Any())` instead of `if (input.Count() == 0)` for the exception above the local function.

IEvangelist reviewed (commented):
This is awesome! Well done.

jmarkman committed:
Address feedback provided in PR

Fixed compile time error in YieldReturn example:
added usings for System.Collections.Generic and System.Linq

Change YieldReturn example to throw an ArgumentException
instead of a base Exception (following best practices) and combine
ToLower on item and yield return into one line instead of two operations

Address written content suggestions/alterations from
@BillWagner and @IEvangelist

jmarkman committed:
Merge branch 'master' of https://github.com/dotnet/docs into jmarkman-organize-local-funcs-page

jmarkman committed:
Extra curly brace was causing build to break

jmarkman commented:
I've gone ahead and made several changes based on feedback and discussion in the PR. I'm a bit more confident about the `Definite assignment` section and onward, but please double-check me on how I've massaged that section based on the discussion.

IEvangelist commented on docs/csharp/programming-guide/classes-and-structs/snippets/local-functions/Program.cs:
```diff
@@ -121,5 +123,25 @@ private async Task<int> SecondStep(int index, string name)
+        //<YieldReturn>
+        public IEnumerable<string> SequenceToLowercase(IEnumerable<string> input)
+        {
+            if (input.Any())
```
```suggestion
            if (!input.Any())
```

This needs to be a negation, otherwise it would throw an exception when there are any items in the `input`.

jmarkman replied:
😳 100% right, this was an oversight on my end

IEvangelist reviewed (changes requested):
We need to fix the bug in code

BillWagner commented on docs/csharp/programming-guide/classes-and-structs/local-functions.md:
```diff
@@ -127,9 +127,11 @@ int M()
 The compiler can determine that `LocalFunction` definitely assigns `y` when called. Because `LocalFunction` is called before the `return` statement, `y` is definitely assigned at the `return` statement.
 
-### Heap Allocations
+Note that when a local function captures variables in the enclosing scope, the local function is implemented as a delegate type in the eyes of the compiler.
```
I suggest dropping "in the eyes of the compiler":

```suggestion
Note that when a local function captures variables in the enclosing scope, the local function is implemented as a delegate type.
```

BillWagner reviewed (commented):
This is great @jmarkman 

I had one small suggestion in addition to @IEvangelist 's comment on the `Any` call. Once you take those suggestions, we'll approve and this is ready to merge.

Thanks again for all the work on this PR. It's been a huge improvement.

jmarkman committed:
Further changes based on PR discussion

Removed extraenous part of sentence regarding
the compiler from 'Heap allocations' section

Add previously missing negation to 'if (input.Any())' in
yield return example

jmarkman committed:
Merge branch 'master' of https://github.com/dotnet/docs into jmarkman-organize-local-funcs-page

BillWagner commented:
Thanks @jmarkman 

I've approved these changes.

@IEvangelist Can you approve as well, and then merge this?
